theano/tensor/basic.py

```diff
@@ -3965,18 +3970,27 @@ def _make_node_internal(self, axis, tensors,
     def perform(self, node, axis_and_tensors, out_):
         out, = out_
         axis, tensors = axis_and_tensors[0], axis_and_tensors[1:]
-        ndim = tensors[0].ndim
-        if axis < -ndim:
-            raise IndexError("Join axis %d out of bounds [0, %d)" %
-                             (axis, ndim))
-
-        out[0] = theano._asarray(numpy.concatenate(tensors, axis=axis),
-                                 dtype=node.outputs[0].type.dtype)
+        nonempty_tensors = [0 if tensor.size == 0 else 1 for tensor in tensors]
+        num_nonempty = numpy.sum(nonempty_tensors)
+        # checking if more than one non-empty tensors are joined.
+        if num_nonempty > 1:
```

Comment:
I would go in the reverse direction for the test.

if num_empty == 1 and the non empty one is the self.view_map (so currently the inputs 0):
   do the new logic:

else:

    do the old logic.